sdks/java/container/agent/build.gradle

```diff
@@ -59,6 +59,7 @@ if (project.hasProperty('java11Home')) {
 project.tasks.each {
     it.onlyIf {
         project.hasProperty('java11Home') || project.hasProperty('java17Home')
-                || JavaVersion.VERSION_1_8.compareTo(JavaVersion.current()) < 0
+                || (JavaVersion.VERSION_1_8.compareTo(JavaVersion.current()) < 0
+                && JavaVersion.VERSION_1_8.compareTo(JavaVersion.toVersion(project.javaVersion)) < 0)
```

Comment:
Without that, we end up trying to compile this with `--release 8` when running on JDK 11, and it fails because it references Java 11-only APIs like `Module`:

```
JAVA_HOME=${JAVA11_HOME?} ./gradlew compileJava
...
> Task :sdks:java:container:agent:compileJava FAILED
/usr/local/google/home/cushon/src/beam/sdks/java/container/agent/src/main/java/org/apache/beam/agent/OpenModuleAgent.java:33: error: cannot find symbol
    Set<Module> modulesToOpen = new HashSet<>();
        ^
  symbol:   class Module
  location: class OpenModuleAgent
```

Reply:
I think the issue is that `applyJavaNature` is occurring before:
```
if (project.hasProperty('java11Home')) {
    javaVersion = "1.11"
    def java11Home = project.findProperty('java11Home')
    project.tasks.withType(JavaCompile) {
        options.fork = true
        options.forkOptions.javaHome = java11Home as File
        options.compilerArgs += ['-Xlint:-path']
    }
} else if (project.hasProperty('java17Home')) {
    javaVersion = "1.17"
    project.tasks.withType(JavaCompile) {
        setJava17Options(options)
        checkerFramework {
            skipCheckerFramework = true
        }
    }
}
```
which means that `javaVersion = 8` and hence `--release 8` is being added.

Moving the `if block` to before `applyJavaVersion` would resolve this.

Reply:
Done, thanks.

When I was testing earlier I hadn't been setting `-Pjava11Home=`, I was just setting `JAVA_HOME` to JDK 11. Is that not a supported configuration? I wasn't sure why it was relying on that vs. checking `JavaVersion.current()`.

Reply:
Lack of familiarity with gradle led to solutions that worked but weren't great.